File: src/contests/cg/puzzle/go/mirrors.go
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(reader, &n)
	r := make([]float64, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &r[i])
	}
	reflected := 0.0
	for i := n - 1; i >= 0; i-- {
		ri := r[i]
		denom := 1 - ri*reflected
		if denom == 0 {
			reflected = ri
		} else {
			reflected = ri + ((1-ri)*(1-ri)*reflected)/denom
		}
	}
	fmt.Printf("%.4f\n", reflected)
}

File: src/contests/cg/puzzle/php/mirrors.php
<?php
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

fgets(STDIN);
$r = array_map('floatval', explode(' ', trim(fgets(STDIN))));
$reflected = 0.0;
for ($i = count($r) - 1; $i >= 0; $i--) {
    $ri = $r[$i];
    $denom = 1 - $ri * $reflected;
    $reflected = $ri + ($denom == 0 ? 0 : ((1 - $ri) * (1 - $ri) * $reflected) / $denom);
}
echo sprintf("%.4f", $reflected) . "\n";
